### PR TITLE
$89 damage effect

### PR DESCRIPTION
**Pull request by MrJackey**
## Summary
Fixes $89
Created blood damage effect.
Spikes now create blood damage effect on player hit.

## Visualisation
coDeq2hXM0

### CODE REVIEW

File: Assets/Scripts/Common/DamageArea.cs [hunk 1]
```diff
@@ -37,6 +44,24 @@ private enum DamageMode {
...
+			foreach (ContactPoint2D contact in contacts) {
+				if (!contact.collider || contact.collider.gameObject != gameObject) continue;
+
+				xValue += contact.point.x;
+				yValue += contact.point.y;
+				i++;
+			}
+
+			if (i > 0)
+				onDamageEffect.Invoke(new Vector2(xValue / i, yValue / i));
```
Comment (smeas) [inline Assets/Scripts/Common/DamageArea.cs]: And maybe extract this into a separate method. `DoEffect` or something?
```suggestion
			if (!other.attachedRigidbody.CompareTag(effectTarget)) return;

			int contactCount = other.GetContacts(contacts);
			if (contactCount != 0) {
				Vector2 averageContact = new Vector2();
				int num = 0;

				for (int i = 0; i < contactCount; i++) {
					ContactPoint2D contact = contacts[i];
					if (contact.collider.gameObject != gameObject) continue;

					averageContact += contact.point;
					num++;
				}

				averageContact /= num;
				onDamageEffect.Invoke(averageContact);
			}
```

File: Assets/Scripts/Common/Effects/DamageEffect.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class DamageEffect : MonoBehaviour {
+	[SerializeField] private ParticleSystem damageEffect;
+
+	private void Start() {
+		GetComponent<DamageArea>().OnDamageEffect.AddListener(Spawn);
```
Comment (smeas) [inline Assets/Scripts/Common/Effects/DamageEffect.cs:10]: Will likely never be a problem, but the fact that this listener is never removed makes me anxious.

File: Assets/Scripts/Common/Effects/ParticleCleanUp.cs [hunk 3]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class ParticleCleanUp : MonoBehaviour {
+	private new ParticleSystem particleSystem;
```
Comment (smeas) [inline Assets/Scripts/Common/Effects/ParticleCleanUp.cs]: Why is this even a field?

**Review by smeas — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Common/DamageArea.cs [hunk 4]
```diff
@@ -45,23 +45,26 @@ private enum DamageMode {
...
-				i++;
-			}
+		foreach (ContactPoint2D contact in contacts) {
+			if (!contact.collider || contact.collider.gameObject != gameObject) continue;
 
-			if (i > 0)
-				onDamageEffect.Invoke(new Vector2(xValue / i, yValue / i));
+			averageContact += contact.point;
+			num++;
 		}
```
Comment (smeas) [inline Assets/Scripts/Common/DamageArea.cs]: You're looping through all 10 items in the array regardless of `contactCount`. Old entries in the array will still be considered and can affect the calculation.

**Review by smeas — CHANGES_REQUESTED** [review]

**Review by smeas — APPROVED** [review]
Cool!

**Review by Addehman — APPROVED** [review]
I bloody approve!

**Pull request merged**